// @grafana/plugin-e2e/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.selectors = exports.expect = exports.test = exports.resolveSelectors = exports.AppPage = exports.PluginConfigPage = exports.AppConfigPage = exports.VariablePage = exports.VariableEditPage = exports.PanelEditPage = exports.GrafanaPage = exports.ExplorePage = exports.DataSourceConfigPage = exports.DashboardPage = exports.AnnotationPage = exports.AnnotationEditPage = exports.TimeRange = exports.Panel = exports.DataSourcePicker = void 0;
const test_1 = require("@playwright/test");
const annotationEditPage_1 = require("./fixtures/annotationEditPage");
const grafanaAPIClient_1 = require("./fixtures/grafanaAPIClient");
const createDataSource_1 = require("./fixtures/commands/createDataSource");
const createDataSourceConfigPage_1 = require("./fixtures/commands/createDataSourceConfigPage");
const createUser_1 = require("./fixtures/commands/createUser");
const gotoAnnotationEditPage_1 = require("./fixtures/commands/gotoAnnotationEditPage");
const gotoAppConfigPage_1 = require("./fixtures/commands/gotoAppConfigPage");
const gotoAppPage_1 = require("./fixtures/commands/gotoAppPage");
const gotoDashboardPage_1 = require("./fixtures/commands/gotoDashboardPage");
const gotoDataSourceConfigPage_1 = require("./fixtures/commands/gotoDataSourceConfigPage");
const gotoPanelEditPage_1 = require("./fixtures/commands/gotoPanelEditPage");
const gotoAlertRuleEditPage_1 = require("./fixtures/commands/gotoAlertRuleEditPage");
const gotoVariableEditPage_1 = require("./fixtures/commands/gotoVariableEditPage");
const login_1 = require("./fixtures/commands/login");
const readProvisionedDashboard_1 = require("./fixtures/commands/readProvisionedDashboard");
const readProvisionedDataSource_1 = require("./fixtures/commands/readProvisionedDataSource");
const readProvisionedAlertRule_1 = require("./fixtures/commands/readProvisionedAlertRule");
const dashboardPage_1 = require("./fixtures/dashboardPage");
const explorePage_1 = require("./fixtures/explorePage");
const grafanaVersion_1 = require("./fixtures/grafanaVersion");
const isFeatureToggleEnabled_1 = require("./fixtures/isFeatureToggleEnabled");
const page_1 = require("./fixtures/page");
const panelEditPage_1 = require("./fixtures/panelEditPage");
const selectors_1 = require("./fixtures/selectors");
const variableEditPage_1 = require("./fixtures/variableEditPage");
const alertRuleEditPage_1 = require("./fixtures/alertRuleEditPage");
const options_1 = require("./options");
const toHaveAlert_1 = require("./matchers/toHaveAlert");
const toDisplayPreviews_1 = require("./matchers/toDisplayPreviews");
const toBeOK_1 = require("./matchers/toBeOK");
const variablePage_1 = require("./fixtures/variablePage");
const gotoVariablePage_1 = require("./fixtures/commands/gotoVariablePage");
// models
var DataSourcePicker_1 = require("./models/components/DataSourcePicker");
Object.defineProperty(exports, "DataSourcePicker", { enumerable: true, get: function () { return DataSourcePicker_1.DataSourcePicker; } });
var Panel_1 = require("./models/components/Panel");
Object.defineProperty(exports, "Panel", { enumerable: true, get: function () { return Panel_1.Panel; } });
var TimeRange_1 = require("./models/components/TimeRange");
Object.defineProperty(exports, "TimeRange", { enumerable: true, get: function () { return TimeRange_1.TimeRange; } });
var AnnotationEditPage_1 = require("./models/pages/AnnotationEditPage");
Object.defineProperty(exports, "AnnotationEditPage", { enumerable: true, get: function () { return AnnotationEditPage_1.AnnotationEditPage; } });
var AnnotationPage_1 = require("./models/pages/AnnotationPage");
Object.defineProperty(exports, "AnnotationPage", { enumerable: true, get: function () { return AnnotationPage_1.AnnotationPage; } });
var DashboardPage_1 = require("./models/pages/DashboardPage");
Object.defineProperty(exports, "DashboardPage", { enumerable: true, get: function () { return DashboardPage_1.DashboardPage; } });
var DataSourceConfigPage_1 = require("./models/pages/DataSourceConfigPage");
Object.defineProperty(exports, "DataSourceConfigPage", { enumerable: true, get: function () { return DataSourceConfigPage_1.DataSourceConfigPage; } });
var ExplorePage_1 = require("./models/pages/ExplorePage");
Object.defineProperty(exports, "ExplorePage", { enumerable: true, get: function () { return ExplorePage_1.ExplorePage; } });
var GrafanaPage_1 = require("./models/pages/GrafanaPage");
Object.defineProperty(exports, "GrafanaPage", { enumerable: true, get: function () { return GrafanaPage_1.GrafanaPage; } });
var PanelEditPage_1 = require("./models/pages/PanelEditPage");
Object.defineProperty(exports, "PanelEditPage", { enumerable: true, get: function () { return PanelEditPage_1.PanelEditPage; } });
var VariableEditPage_1 = require("./models/pages/VariableEditPage");
Object.defineProperty(exports, "VariableEditPage", { enumerable: true, get: function () { return VariableEditPage_1.VariableEditPage; } });
var VariablePage_1 = require("./models/pages/VariablePage");
Object.defineProperty(exports, "VariablePage", { enumerable: true, get: function () { return VariablePage_1.VariablePage; } });
var AppConfigPage_1 = require("./models/pages/AppConfigPage");
Object.defineProperty(exports, "AppConfigPage", { enumerable: true, get: function () { return AppConfigPage_1.AppConfigPage; } });
var PluginConfigPage_1 = require("./models/pages/PluginConfigPage");
Object.defineProperty(exports, "PluginConfigPage", { enumerable: true, get: function () { return PluginConfigPage_1.PluginConfigPage; } });
var AppPage_1 = require("./models/pages/AppPage");
Object.defineProperty(exports, "AppPage", { enumerable: true, get: function () { return AppPage_1.AppPage; } });
var resolver_1 = require("./e2e-selectors/resolver");
Object.defineProperty(exports, "resolveSelectors", { enumerable: true, get: function () { return resolver_1.resolveSelectors; } });
// types
__exportStar(require("./types"), exports);
exports.test = test_1.test.extend({
    selectors: selectors_1.selectors,
    grafanaVersion: grafanaVersion_1.grafanaVersion,
    login: login_1.login,
    grafanaAPIClient: grafanaAPIClient_1.grafanaAPIClient,
    createDataSourceConfigPage: createDataSourceConfigPage_1.createDataSourceConfigPage,
    page: page_1.page,
    dashboardPage: dashboardPage_1.dashboardPage,
    panelEditPage: panelEditPage_1.panelEditPage,
    variableEditPage: variableEditPage_1.variableEditPage,
    variablePage: variablePage_1.variablePage,
    annotationEditPage: annotationEditPage_1.annotationEditPage,
    alertRuleEditPage: alertRuleEditPage_1.alertRuleEditPage,
    explorePage: explorePage_1.explorePage,
    createDataSource: createDataSource_1.createDataSource,
    readProvisionedDataSource: readProvisionedDataSource_1.readProvisionedDataSource,
    readProvisionedAlertRule: readProvisionedAlertRule_1.readProvisionedAlertRule,
    readProvisionedDashboard: readProvisionedDashboard_1.readProvisionedDashboard,
    isFeatureToggleEnabled: isFeatureToggleEnabled_1.isFeatureToggleEnabled,
    createUser: createUser_1.createUser,
    gotoDashboardPage: gotoDashboardPage_1.gotoDashboardPage,
    gotoPanelEditPage: gotoPanelEditPage_1.gotoPanelEditPage,
    gotoVariableEditPage: gotoVariableEditPage_1.gotoVariableEditPage,
    gotoVariablePage: gotoVariablePage_1.gotoVariablePage,
    gotoAnnotationEditPage: gotoAnnotationEditPage_1.gotoAnnotationEditPage,
    gotoAlertRuleEditPage: gotoAlertRuleEditPage_1.gotoAlertRuleEditPage,
    gotoDataSourceConfigPage: gotoDataSourceConfigPage_1.gotoDataSourceConfigPage,
    gotoAppConfigPage: gotoAppConfigPage_1.gotoAppConfigPage,
    gotoAppPage: gotoAppPage_1.gotoAppPage,
    ...options_1.options,
});
exports.expect = test_1.expect.extend({
    toHaveAlert: toHaveAlert_1.toHaveAlert,
    toDisplayPreviews: toDisplayPreviews_1.toDisplayPreviews,
    toBeOK: toBeOK_1.toBeOK,
});
var test_2 = require("@playwright/test");
Object.defineProperty(exports, "selectors", { enumerable: true, get: function () { return test_2.selectors; } });
