// @grafana/plugin-e2e/dist/e2e-selectors/versioned/apis.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.versionedAPIs = void 0;
const constants_1 = require("./constants");
exports.versionedAPIs = {
    Alerting: {
        eval: {
            [constants_1.MIN_GRAFANA_VERSION]: '/api/v1/eval',
        },
    },
    DataSource: {
        resourcePattern: {
            [constants_1.MIN_GRAFANA_VERSION]: '/api/datasources/*/resources',
        },
        resourceUIDPattern: {
            '9.4.4': '/api/datasources/uid/*/resources',
            [constants_1.MIN_GRAFANA_VERSION]: '/api/datasources/*/resources',
        },
        queryPattern: {
            [constants_1.MIN_GRAFANA_VERSION]: '*/**/api/ds/query*',
        },
        query: {
            [constants_1.MIN_GRAFANA_VERSION]: '/api/ds/query',
        },
        health: {
            ['9.5.0']: (uid, _) => `/api/datasources/uid/${uid}/health`,
            [constants_1.MIN_GRAFANA_VERSION]: (_, id) => `/api/datasources/${id}/health`,
        },
        datasourceByUID: {
            [constants_1.MIN_GRAFANA_VERSION]: (uid) => `/api/datasources/uid/${uid}`,
        },
        proxy: {
            '9.4.0': (uid, _) => `api/datasources/proxy/uid/${uid}`,
            [constants_1.MIN_GRAFANA_VERSION]: (_, id) => `/api/datasources/proxy/${id}`,
        },
    },
    Dashboard: {
        delete: {
            [constants_1.MIN_GRAFANA_VERSION]: (uid) => `/api/dashboards/uid/${uid}`,
        },
    },
    Plugin: {
        settings: {
            [constants_1.MIN_GRAFANA_VERSION]: (pluginId) => `/api/plugins/${pluginId}/settings`,
        },
    },
};
