// @grafana/plugin-e2e/dist/e2e-selectors/versioned/components.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.versionedComponents = void 0;
const constants_1 = require("./constants");
exports.versionedComponents = {
    TimePicker: {
        openButton: {
            '8.1.0': 'data-testid TimePicker Open Button',
            [constants_1.MIN_GRAFANA_VERSION]: 'TimePicker open button',
        },
        fromField: {
            '10.2.3': 'data-testid Time Range from field',
            [constants_1.MIN_GRAFANA_VERSION]: 'Time Range from field',
        },
        toField: {
            '10.2.3': 'data-testid Time Range to field',
            [constants_1.MIN_GRAFANA_VERSION]: 'Time Range to field',
        },
        applyTimeRange: {
            '8.1.0': 'data-testid TimePicker submit button',
            [constants_1.MIN_GRAFANA_VERSION]: 'TimePicker submit button',
        },
        absoluteTimeRangeTitle: {
            [constants_1.MIN_GRAFANA_VERSION]: 'data-testid-absolute-time-range-narrow',
        },
    },
    Menu: {
        MenuComponent: {
            [constants_1.MIN_GRAFANA_VERSION]: (title) => `${title} menu`,
        },
        MenuGroup: {
            [constants_1.MIN_GRAFANA_VERSION]: (title) => `${title} menu group`,
        },
        MenuItem: { [constants_1.MIN_GRAFANA_VERSION]: (title) => `${title} menu item` },
        SubMenu: {
            container: {
                '10.3.0': 'data-testid SubMenu container',
                [constants_1.MIN_GRAFANA_VERSION]: 'SubMenu',
            },
            icon: {
                '10.3.0': 'data-testid SubMenu icon',
                [constants_1.MIN_GRAFANA_VERSION]: 'SubMenu icon',
            },
        },
    },
    Panels: {
        Panel: {
            title: {
                '8.1.2': (title) => `data-testid Panel header ${title}`,
                [constants_1.MIN_GRAFANA_VERSION]: (title) => `Panel header ${title}`,
            },
            headerCornerInfo: {
                [constants_1.MIN_GRAFANA_VERSION]: (mode) => `Panel header ${mode}`,
            },
            status: {
                ['10.2.0']: (status) => `data-testid Panel status ${status}`,
                [constants_1.MIN_GRAFANA_VERSION]: (_) => 'Panel status',
            },
            toggleTableViewPanel: {
                '10.4.0': (title) => `data-testid Panel header ${title}`,
                [constants_1.MIN_GRAFANA_VERSION]: (_) => 'data-testid Panel',
            },
            PanelDataErrorMessage: {
                '10.4.0': 'data-testid Panel data error message',
            },
            menuItems: { '9.5.0': (item) => `data-testid Panel menu item ${item}` },
            menu: { '9.5.0': (item) => `data-testid Panel menu ${item}` },
        },
        Visualization: {
            Table: {
                header: {
                    [constants_1.MIN_GRAFANA_VERSION]: 'table header',
                },
                footer: {
                    [constants_1.MIN_GRAFANA_VERSION]: 'table-footer',
                },
                body: {
                    // did not exist prior to 10.2.0
                    '10.2.0': 'data-testid table body',
                },
            },
        },
    },
    VizLegend: {
        seriesName: {
            [constants_1.MIN_GRAFANA_VERSION]: (name) => `VizLegend series ${name}`,
        },
    },
    Drawer: {
        General: {
            title: {
                [constants_1.MIN_GRAFANA_VERSION]: (title) => `Drawer title ${title}`,
            },
        },
    },
    PanelEditor: {
        General: {
            content: {
                '11.1.0': 'data-testid Panel editor content',
                [constants_1.MIN_GRAFANA_VERSION]: 'Panel editor content',
            },
        },
        applyButton: {
            '9.2.0': 'data-testid Apply changes and go back to dashboard',
            [constants_1.MIN_GRAFANA_VERSION]: 'Apply changes and go back to dashboard',
        },
        toggleVizPicker: {
            '10.0.0': 'data-testid toggle-viz-picker',
            [constants_1.MIN_GRAFANA_VERSION]: 'toggle-viz-picker',
        },
        OptionsPane: {
            content: {
                '11.1.0': 'data-testid Panel editor option pane content',
                [constants_1.MIN_GRAFANA_VERSION]: 'Panel editor option pane content',
            },
            fieldLabel: {
                [constants_1.MIN_GRAFANA_VERSION]: (type) => `${type} field property editor`,
            },
            fieldInput: {
                '11.0.0': (title) => `data-testid Panel editor option pane field input ${title}`,
            },
        },
    },
    RefreshPicker: {
        runButtonV2: {
            ['8.3.0']: 'data-testid RefreshPicker run button',
            [constants_1.MIN_GRAFANA_VERSION]: 'RefreshPicker run button',
        },
    },
    QueryTab: {
        addQuery: {
            '10.2.0': 'data-testid query-tab-add-query',
            [constants_1.MIN_GRAFANA_VERSION]: 'Query editor add query button',
        },
        addExpression: {
            '11.0.0': 'data-testid query-tab-add-expression',
            '9.5.2': 'query-tab-add-expression',
        },
    },
    QueryEditorRows: {
        rows: {
            [constants_1.MIN_GRAFANA_VERSION]: 'Query editor row',
        },
    },
    QueryEditorRow: {
        title: {
            [constants_1.MIN_GRAFANA_VERSION]: (refId) => `Query editor row title ${refId}`,
        },
        actionButton: {
            '10.4.0': (title) => `data-testid ${title}`,
            [constants_1.MIN_GRAFANA_VERSION]: (title) => `${title}`,
        },
    },
    AlertRules: {
        previewButton: { '11.1.0': 'data-testid alert-rule preview-button' },
        ruleNameField: { '11.1.0': 'data-testid alert-rule name-field' },
        newFolderButton: { '11.1.0': 'data-testid alert-rule new-folder-button' },
        newFolderNameField: { '11.1.0': 'data-testid alert-rule name-folder-name-field' },
        newFolderNameCreateButton: { '11.1.0': 'data-testid alert-rule name-folder-name-create-button' },
        newEvaluationGroupButton: { '11.1.0': 'data-testid alert-rule new-evaluation-group-button' },
        newEvaluationGroupName: { '11.1.0': 'data-testid alert-rule new-evaluation-group-name' },
        newEvaluationGroupInterval: { '11.1.0': 'data-testid alert-rule new-evaluation-group-interval' },
        newEvaluationGroupCreate: { '11.1.0': 'data-testid alert-rule new-evaluation-group-create-button' },
    },
    Alert: {
        alertV2: {
            '8.3.0': (severity) => `data-testid Alert ${severity}`,
            [constants_1.MIN_GRAFANA_VERSION]: (severity) => `Alert ${severity}`,
        },
    },
    NavToolbar: {
        editDashboard: {
            editButton: {
                '11.1.0': 'data-testid Edit dashboard button',
            },
            backToDashboardButton: {
                '11.1.0': 'data-testid Back to dashboard button',
            },
        },
    },
    PageToolbar: {
        item: {
            [constants_1.MIN_GRAFANA_VERSION]: (tooltip) => `${tooltip}`,
        },
        showMoreItems: {
            [constants_1.MIN_GRAFANA_VERSION]: 'Show more items',
        },
        itemButton: {
            //did not exist prior to 9.5.0
            ['9.5.0']: (title) => `data-testid ${title}`,
        },
        itemButtonTitle: {
            '10.1.0': 'Add button',
            [constants_1.MIN_GRAFANA_VERSION]: 'Add panel button',
        },
    },
    QueryEditorToolbarItem: {
        button: {
            [constants_1.MIN_GRAFANA_VERSION]: (title) => `QueryEditor toolbar item button ${title}`,
        },
    },
    OptionsGroup: {
        group: {
            '11.1.0': (title) => (title ? `data-testid Options group ${title}` : 'data-testid Options group'),
            [constants_1.MIN_GRAFANA_VERSION]: (title) => (title ? `Options group ${title}` : 'Options group'),
        },
        toggle: {
            '11.1.0': (title) => title ? `data-testid Options group ${title} toggle` : 'data-testid Options group toggle',
            [constants_1.MIN_GRAFANA_VERSION]: (title) => (title ? `Options group ${title} toggle` : 'Options group toggle'),
        },
        groupTitle: {
            [constants_1.MIN_GRAFANA_VERSION]: 'Panel options',
        },
    },
    PluginVisualization: {
        item: {
            [constants_1.MIN_GRAFANA_VERSION]: (title) => `Plugin visualization item ${title}`,
        },
    },
    Select: {
        option: {
            '11.1.0': 'data-testid Select option',
            [constants_1.MIN_GRAFANA_VERSION]: 'Select option',
        },
        input: {
            '8.3.0': () => 'input[id*="time-options-input"]',
            [constants_1.MIN_GRAFANA_VERSION]: () => 'input[id*="react-select-"]',
        },
        singleValue: {
            [constants_1.MIN_GRAFANA_VERSION]: () => 'div[class*="-singleValue"]',
        },
    },
    DataSourcePicker: {
        container: {
            '10.0.0': 'data-testid Data source picker select container',
            // did not exist prior to 8.3.0
            '8.3.0': 'Data source picker select container',
        },
    },
    TimeZonePicker: {
        containerV2: {
            '8.3.0': 'data-testid Time zone picker select container',
            [constants_1.MIN_GRAFANA_VERSION]: 'Folder picker select container',
        },
        changeTimeSettingsButton: {
            '11.0.0': 'data-testid Time zone picker Change time settings button',
        },
    },
    CodeEditor: {
        container: {
            '10.2.3': 'data-testid Code editor container',
            [constants_1.MIN_GRAFANA_VERSION]: 'Code editor container',
        },
    },
    Annotations: {
        editor: {
            testButton: {
                '11.0.0': 'data-testid annotations-test-button',
            },
            resultContainer: {
                '11.0.0': 'data-testid annotations-query-result-container',
            },
        },
    },
    QueryField: {
        container: {
            '10.3.0': 'data-testid Query field',
            [constants_1.MIN_GRAFANA_VERSION]: 'Query field',
        },
    },
};
