// @grafana/plugin-e2e/dist/e2e-selectors/versioned/pages.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.versionedPages = void 0;
const constants_1 = require("./constants");
exports.versionedPages = {
    Home: {
        url: {
            [constants_1.MIN_GRAFANA_VERSION]: '/',
        },
    },
    Alerting: {
        AddAlertRule: {
            url: {
                ['10.1.0']: '/alerting/new/alerting',
                [constants_1.MIN_GRAFANA_VERSION]: '/alerting/new',
            },
        },
        EditAlertRule: {
            url: {
                [constants_1.MIN_GRAFANA_VERSION]: (alertRuleUid) => `alerting/${alertRuleUid}/edit`,
            },
        },
    },
    DataSource: {
        saveAndTest: {
            '10.0.0': 'data-testid Data source settings page Save and Test button',
            [constants_1.MIN_GRAFANA_VERSION]: 'Data source settings page Save and Test button',
        },
    },
    EditDataSource: {
        url: {
            '10.2.0': (dataSourceUid) => `/connections/datasources/edit/${dataSourceUid}`,
            [constants_1.MIN_GRAFANA_VERSION]: (dataSourceUid) => `/datasources/edit/${dataSourceUid}`,
        },
    },
    AddDashboard: {
        url: {
            [constants_1.MIN_GRAFANA_VERSION]: '/dashboard/new',
        },
        itemButton: {
            //did not exist prior to 9.5.0
            '9.5.0': (title) => `data-testid ${title}`,
        },
        addNewPanel: {
            '11.1.0': 'data-testid Add new panel',
            [constants_1.MIN_GRAFANA_VERSION]: 'Add new panel',
        },
        itemButtonAddViz: {
            [constants_1.MIN_GRAFANA_VERSION]: 'Add new visualization menu item',
        },
        Settings: {
            Annotations: {
                List: {
                    url: {
                        [constants_1.MIN_GRAFANA_VERSION]: '/dashboard/new?orgId=1&editview=annotations',
                    },
                },
                Edit: {
                    url: {
                        [constants_1.MIN_GRAFANA_VERSION]: (annotationIndex) => `/dashboard/new?editview=annotations&editIndex=${annotationIndex}`,
                    },
                },
            },
            Variables: {
                List: {
                    url: {
                        [constants_1.MIN_GRAFANA_VERSION]: '/dashboard/new?orgId=1&editview=templating',
                        '11.3.0': '/dashboard/new?orgId=1&editview=variables',
                    },
                },
                Edit: {
                    url: {
                        [constants_1.MIN_GRAFANA_VERSION]: (variableIndex) => `/dashboard/new?orgId=1&editview=templating&editIndex=${variableIndex}`,
                        '11.3.0': (variableIndex) => `/dashboard/new?orgId=1&editview=variables&editIndex=${variableIndex}`,
                    },
                },
            },
        },
    },
    Dashboard: {
        url: {
            [constants_1.MIN_GRAFANA_VERSION]: (uid) => `/d/${uid}`,
        },
        Settings: {
            Actions: {
                close: {
                    [constants_1.MIN_GRAFANA_VERSION]: 'Go Back button',
                    '9.5.0': 'data-testid dashboard-settings-close',
                },
            },
            Annotations: {
                Edit: {
                    url: {
                        [constants_1.MIN_GRAFANA_VERSION]: (dashboardUid, annotationIndex) => `/d/${dashboardUid}?editview=annotations&editIndex=${annotationIndex}`,
                    },
                },
                List: {
                    url: {
                        [constants_1.MIN_GRAFANA_VERSION]: (dashboardUid) => `/d/${dashboardUid}?editview=annotations`,
                    },
                    addAnnotationCTA: 'Call to action button Add annotation query',
                    addAnnotationCTAV2: {
                        //did not exist prior to 8.3.0
                        '8.3.0': 'data-testid Call to action button Add annotation query',
                    },
                },
            },
            Variables: {
                List: {
                    url: {
                        [constants_1.MIN_GRAFANA_VERSION]: (dashboardUid) => `/d/${dashboardUid}?editview=templating`,
                        '11.3.0': (dashboardUid) => `/d/${dashboardUid}?editview=variables`,
                    },
                    newButton: {
                        [constants_1.MIN_GRAFANA_VERSION]: 'Variable editor New variable button',
                    },
                    table: {
                        [constants_1.MIN_GRAFANA_VERSION]: 'Variable editor Table',
                    },
                    addVariableCTAV2: {
                        [constants_1.MIN_GRAFANA_VERSION]: (name) => `data-testid Call to action button ${name}`,
                    },
                    addVariableCTAV2Item: {
                        [constants_1.MIN_GRAFANA_VERSION]: 'Add variable',
                    },
                },
                Edit: {
                    url: {
                        [constants_1.MIN_GRAFANA_VERSION]: (dashboardUid, editIndex) => `/d/${dashboardUid}?editview=templating&editIndex=${editIndex}`,
                        '11.3.0': (dashboardUid, editIndex) => `/d/${dashboardUid}?editview=variables&editIndex=${editIndex}`,
                    },
                    General: {
                        generalTypeSelectV2: {
                            '8.5.0': 'data-testid Variable editor Form Type select',
                            [constants_1.MIN_GRAFANA_VERSION]: 'Variable editor Form Type select',
                        },
                        previewOfValuesOption: {
                            '10.4.0': 'data-testid Variable editor Preview of Values option',
                            [constants_1.MIN_GRAFANA_VERSION]: 'Variable editor Preview of Values option',
                        },
                        submitButton: {
                            '10.4.0': 'data-testid Variable editor Run Query button',
                            [constants_1.MIN_GRAFANA_VERSION]: 'Variable editor Submit button',
                        },
                        selectionOptionsIncludeAllSwitch: {
                            [constants_1.MIN_GRAFANA_VERSION]: 'Variable editor Form IncludeAll switch',
                        },
                        generalNameInputV2: {
                            '8.5.0': 'data-testid Variable editor Form Name field',
                            [constants_1.MIN_GRAFANA_VERSION]: 'Variable editor Form Name field',
                        },
                        applyButton: {
                            '9.3.0': 'data-testid Variable editor Apply button',
                        },
                    },
                },
            },
        },
        SubMenu: {
            submenuItemLabels: {
                [constants_1.MIN_GRAFANA_VERSION]: (item) => `data-testid Dashboard template variables submenu Label ${item}`,
            },
            submenuItemValueDropDownValueLinkTexts: {
                [constants_1.MIN_GRAFANA_VERSION]: (item) => `data-testid Dashboard template variables Variable Value DropDown value link text ${item}`,
            },
            submenuItemValueDropDownDropDown: { [constants_1.MIN_GRAFANA_VERSION]: 'Variable options' },
            submenuItemValueDropDownOptionTexts: {
                [constants_1.MIN_GRAFANA_VERSION]: (item) => `data-testid Dashboard template variables Variable Value DropDown option text ${item}`,
            },
        },
        SaveDashboardAsModal: {
            newName: {
                [constants_1.MIN_GRAFANA_VERSION]: 'Save dashboard title field',
            },
            save: {
                [constants_1.MIN_GRAFANA_VERSION]: 'Save dashboard button',
            },
        },
    },
    Explore: {
        url: {
            [constants_1.MIN_GRAFANA_VERSION]: '/explore',
        },
    },
    Plugin: {
        url: {
            [constants_1.MIN_GRAFANA_VERSION]: (pluginId) => `/plugins/${pluginId}`,
        },
    },
};
