// @grafana/plugin-e2e/dist/matchers/toHaveAlert.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.toHaveAlert = void 0;
const test_1 = require("@playwright/test");
const utils_1 = require("./utils");
const toHaveAlert = async (grafanaPage, severity, options) => {
    let pass = true;
    let message = `An alert of variant ${severity} to be displayed on the page`;
    try {
        const filteredAlerts = grafanaPage
            .getByGrafanaSelector(grafanaPage.ctx.selectors.components.Alert.alertV2(severity))
            .filter({
            hasText: options?.hasText,
            hasNotText: options?.hasNotText,
            has: options?.has,
            hasNot: options?.hasNot,
        });
        await (0, test_1.expect)(filteredAlerts.first()).toBeVisible({ timeout: options?.timeout });
    }
    catch (err) {
        message = (0, utils_1.getMessage)(message, `No alert was found on the page: ${err instanceof Error ? err.toString() : ''}`);
        pass = false;
    }
    return {
        message: () => message,
        pass,
    };
};
exports.toHaveAlert = toHaveAlert;
